bin/nodetool-wrapper

```
@@ -35,6 +35,47 @@ function get_help_opt() {
     echo "$help_opt"
 }
 
+function get_api_port() {
+    local api_port=10000
+    for opt in "$@"; do
+        # if -Dcom.scylladb.apiPort=<num> exists in the command line options,
+        # override the "-p" option with "<num>"
+        if [[ "$opt" =~ ^-Dcom.scylladb.apiPort= ]]; then
+            api_port=${opt#*=}
+        fi
+    done
+    echo "$api_port"
+}
+
+function override_port_option() {
+    echo "$@"
+    local api_port=$1
+    shift
+    local new_args=()
+    # override the -p or --port option with the $api_port
+    # api_port is 10000 by default
+
+    while [ "$#" -gt 0 ]; do
+        case "$1" in
+            -Dcom.scylladb.apiPort=*)
+                # swallow it
+                shift
+                ;;
+            -p|--port)
+                # override the argument
+                new_args+=("$1")
+                new_args+=("$api_port")
+                shift 2
```

Do both the original and new nodetool support a -p option? That's wrong since they refer to different ports.

Better have a new option for the REST port.

---

> Do both the original and new nodetool support a -p option? That's wrong since they refer to different ports.

yes, they both support 

- `-p <port>`, `--port=<port>` and `--port <port>`, 
- `-h <host>`, `--host=<host>` and `--host <host>`

> 
> Better have a new option for the REST port.

IMHO, from the `scylla nodetool` 's perspective, `-p` and `--port` is a natural choice. as it just talks to the API port. as an alternative, probably we could add a new option to the nodetool-wrapper? but that's exactly what `-Dcom.scylladb.apiPort` does. i didn't add a proper option because i didn't wanted

1. filter the new option for the cassandra nodetool tool
2. translate it to `-p` for "scylla nodetool"
3. remove the `-p` for "scylla nodetool"

with `-Dcom.scylladb.apiPort`, i just need to 

1. override it to `-p` for "scylla nodetool"

---

I would prefer if any hacks around the transition from the legacy nodetool to the native nodetool, would be concentrated in this file as much as possible. So that when we are done with the transition we can just drop this file and drop all the hacks together with it.

Thinking ahead, having a single `-p|--port` option for the native nodetool is preferable, then having an `--api-port` and a dead `-p|--port`, which is silently ignored because it used to refer to the now defunct JMX port.

IMO we should strive to have a clear interface long-term and tolerate some confusion during the transition if we have to.

---

(the above is to say I agree with the current approach and I prefer it instead of adding a new port option)

---

It looks cleaner but doesn't work. If you had a script that called nodetool with the jmx port, it wouldn't work with native nodetool, and vice versa.

---

@denesb hi Botond, i am convinced by Avi now. we might need to add a `--api-port` to *this script* before fully migrating to "scylla nodetool". by then we can drop this option, well, not exactly, we would be dropping this script. and everything would just work fine.

---

I was actually convinced by Botond ;-) A user reads in some document that you need to pass the port with "-p", and he doesn't care what this port means, whether it is JMX or REST API.
Using this option would be rare anyway, but if a user does need to choose a port explicitly, no need to invent two differently-named options to do it.

---

If we add `--api-port` to this script, we also have to add it to `scylla-nodetool` and also have to teach it to expect both to be set, and which one to choose if so. We can do that to ease the transition.

---

We should change scylla-nodetool to reject --port.

This is similar to accepting a measurement without units. The tools don't need "a port". They need a port that implements a specific protocol, but each wants a different protocol. They're not compatible in that respect, and we shouldn't pretend they are.

Yes, it's ugly if the way to specify scylla-nodetool's port isn't --port, but that's life.

---

By the same vein, we should reject `-h|--host`. And what about the other legacy JMX related parameters which we silently accept (and ignore)? I think adding two more paremeters to the already longish list of parameters we reinterpret/ignore is not breaking anything here.

Our compatibility is pretty shallow after all and so far our policy has been to cause the least surprise possible while not standing in the way. I think rejecting `--host` and `--port` would definitely qualify as "standing in the way".

---

I still think that whatever scripts do with these options is not relevant and not a problem. The only thing I'm thinking about is **documentation** - people will look on Google for nodetool documentation, and hope that whatever examples they see there, just works. If the example tells you "you can use -p to specify a different port", you try that.
But to be honest, I don't think it's very important. If "-p" doesn't work, the user will just use "--help" and expect to see the correct list of options. Just please make sure that "--help" isn't forwarded to the other nodetool :-)